comm/src/main/scala/coop/rchain/comm/rp/HandleMessages.scala

```diff
@@ -74,51 +73,47 @@ object HandleMessages {
       _ <- ConnectionsCell[F].addConnAndReport(peer)
     } yield handledWithoutMessage
 
+  // TODO: investigate what's with unused parameter - protocolHandshake
   def handleProtocolHandshake[F[_]: Monad: TransportLayer: Log: ConnectionsCell: RPConfAsk: Metrics](
       peer: PeerNode,
       protocolHandshake: ProtocolHandshake
   ): F[CommunicationResponse] = {
-    def isValidPublicInetAddress(host: String): Boolean =
+    def isLocalAddress(host: String): Boolean =
       Try(InetAddress.getByName(host))
-        .map { addr =>
-          !(addr.isAnyLocalAddress ||
-            addr.isLinkLocalAddress ||
-            addr.isLoopbackAddress ||
-            addr.isMulticastAddress ||
-            addr.isSiteLocalAddress)
+        .map { a =>
+          a.isAnyLocalAddress || a.isLinkLocalAddress || a.isLoopbackAddress || a.isMulticastAddress || a.isSiteLocalAddress
         }
         .getOrElse(false)
 
-    def canAddNewHost(conf: RPConf): F[Boolean] = {
-      val isCurrentPeerLocal  = isValidPublicInetAddress(conf.local.endpoint.host)
-      val isIncomingPeerLocal = isValidPublicInetAddress(peer.endpoint.host)
-      val bothPeersAreLocal   = isCurrentPeerLocal && isIncomingPeerLocal
-      val bothPeersArePublic  = !isCurrentPeerLocal && !isIncomingPeerLocal
+    def checkPeerOnSameNetwork(conf: RPConf): Boolean = {
+      val isCurrentPeerLocal  = isLocalAddress(conf.local.endpoint.host)
+      val isIncomingPeerLocal = isLocalAddress(peer.endpoint.host)
 
-      (bothPeersAreLocal || bothPeersArePublic).pure[F]
+      // Peer must be in the same subnetwork (in this case public or local) to accept connections
+      isCurrentPeerLocal == isIncomingPeerLocal
     }
 
     def acceptConnection(conf: RPConf): F[CommunicationResponse] = {
       val response = ProtocolHelper.protocolHandshakeResponse(conf.local, conf.networkId)
       for {
         resErr <- TransportLayer[F].send(peer, response)
-        _ <- resErr.fold(
-              kp(().pure[F]),
-              kp(
-                Log[F].info(s"Responded to protocol handshake request from $peer") >>
-                  ConnectionsCell[F].addConnAndReport(peer)
-              )
-            )
+        _ <- resErr.traverse_ { _ =>
+              Log[F].info(s"Responded to protocol handshake request from $peer") *>
+                ConnectionsCell[F].addConnAndReport(peer)
+            }
       } yield handledWithoutMessage
     }
 
+    // TODO: investigate if response should be with an error instead of HandledWitoutMessage
     def declineConnection: F[CommunicationResponse] =
       Log[F]
-        .info(s"Unable to add connection: incoming address ${peer.toAddress}")
+        .info(s"Peer connection rejected, IP not on the same subnetwork, peer: ${peer.toAddress}.")
```

Comment:
IP subnetwork has a well defined meaning, I think it's misleading to use `subnetwork ` in this particular settings. But I cannot figure out better name, so let's keep it. Anyway this is not the full solution.

Reply:
Here are a few examples when for public/private subnets.
- https://docs.aws.amazon.com/vpc/latest/userguide/VPC_Scenario2.html
- https://cloud.google.com/vpc/docs/subnets
- https://sourcedaddy.com/networking/private-subnets.html